### PR TITLE
Changing CiperSuiteId from single to multiple resource

### PR DESCRIPTION
**Pull request by Warmek**
Changing CipherSuiteId field in BootstrapConfig.java from single resource to multiple resource as to match OMA specification.

It aims to implement https://github.com/eclipse/leshan/issues/1402

### CODE REVIEW

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 1]
```diff
@@ -465,9 +480,36 @@ public String toString() {
...
+
+        public CipherSuiteId(byte firstByte, byte secondByte) {
+            this.firstByte = firstByte;
+            this.secondByte = secondByte;
+        }
+
+        /**
+         * The IANA TLS ciphersuite registry is maintained at
+         * https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml. As an example, the
+         * TLS_PSK_WITH_AES_128_CCM_8 ciphersuite is represented with the following string "0xC0,0xA8"
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:496]: I feel this java doc is better to explain constructor above :point_up:

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 2]
```diff
@@ -465,9 +480,36 @@ public String toString() {
...
+         */
+
+        public CipherSuiteId(ULong valueFromSecurityObject) {
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:499]: For this constructor you need to explain(or take the explanation) about how this ULong is created.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 3]
```diff
@@ -465,9 +480,36 @@ public String toString() {
...
+            this.firstByte = firstByte;
+            this.secondByte = secondByte;
+        }
+
+        /**
+         * The IANA TLS ciphersuite registry is maintained at
+         * https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml. As an example, the
+         * TLS_PSK_WITH_AES_128_CCM_8 ciphersuite is represented with the following string "0xC0,0xA8"
+         */
+
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:498]: This empty line should be removed.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 4]
```diff
@@ -362,9 +373,13 @@ public String toString() {
...
 
-        /** The Object ID of the Object Instance for which ACL are applied. */
+        /**
+         * The Object ID of the Object Instance for which ACL are applied.
+         */
         public int objectId;
-        /** The Object instance ID of the Object Instance for which ACL are applied. */
+        /**
+         * The Object instance ID of the Object Instance for which ACL are applied.
+         */
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:382]: any reason to change javadoc format here ?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 5]
```diff
@@ -96,12 +96,24 @@ public class BootstrapConfig {
...
 
-        /** Used as link to associate server Object Instance. */
+        /**
+         * Used as link to associate server Object Instance.
+         */
         public int shortId;
-        /** Specify the lifetime of the registration in seconds (see Section 5.3 Registration). */
+        /**
+         * Specify the lifetime of the registration in seconds (see Section 5.3 Registration).
+         */
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:116]: Any reasons to change javadoc format here ?

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mSingleResource.java [hunk 6]
```diff
@@ -206,23 +210,23 @@ public Object getValue() {
     }
 
     /**
-     * @exception NoSuchElementException use {@link #getValue()} instead.
+     * @throws NoSuchElementException use {@link #getValue()} instead.
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mSingleResource.java:213]: any reason to replace @exception by @throws ?

File: leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mSingleResource.java [hunk 7]
```diff
@@ -181,6 +181,10 @@ public static LwM2mSingleResource newUnsignedIntegerResource(int id, Long value)
             throw new LwM2mNodeException("Invalid value : positive value expected for UNSIGNED_INTEGER");
     }
 
+    public static LwM2mSingleResource newULongArray(int id, ULong[] value) {
+        return new LwM2mSingleResource(id, value, Type.OPAQUE);
+    }
+
```
Comment (sbernard31) [inline leshan-core/src/main/java/org/eclipse/leshan/core/node/LwM2mSingleResource.java:187]: This sounds no more used, should be deleted.

Do not hesitate to review your own pull request, to find this kind of issue by yourself :wink:

**Comment by sbernard31** [comment]
About the title of the PR, better to find a title explaining the content of the PR instead of name of your branch.

About the description, this seems to not really describe the content of the PR (by the way it point to an issue which doesn't exist)

About your git config, you still don't change your  it like asked at (https://github.com/eclipse/leshan/pull/1326#issuecomment-1346841349, https://github.com/eclipse/leshan/pull/1377#issuecomment-1384187634)

Do not hesitate to ask for help (to me or maybe @JaroslawLegierski), if you face any issue.

**Comment by sbernard31** [comment]
This moves forward :slightly_smiling_face:  but still missing some point : 

 - :x:  About the title of the PR, better to find a title explaining the content of the PR instead of name of your branch.
 - :heavy_check_mark:  About the description ...
- :heavy_check_mark:  About your git config  ...
- :x:  missing javadoc on constructor (see review above)

Like I said at https://github.com/eclipse/leshan/issues/1402#issuecomment-1443497522 , maybe javadoc could be a bit improved.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 8]
```diff
@@ -465,6 +456,56 @@ public String toString() {
...
+        /**
+         * Integer is split into 2 bytes for example 49320 (0xc0a8 in hex) will be split into "0xC0,0xA8".
+         * This format is used by Security Object, resource 16.
+         *
+         * @param valueFromSecurityObject Integer representing ciphersuite id
+         */
+        public CipherSuiteId(ULong valueFromSecurityObject) {
+            String binaryString = Long.toBinaryString(valueFromSecurityObject.longValue());
+            this.firstByte = (byte) Integer.parseInt(binaryString.substring(0, 8), 2);
+            this.secondByte = (byte) Integer.parseInt(binaryString.substring(9, 17), 2);
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:486]: Using String is maybe not the best idea.

Should be better to play with binary operator like : 
- shift operator :  `>>`

You can create unit test to check your code.

We should also check that `valueFromSecurityObject` is a 16-bit unsigned integer.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 9]
```diff
@@ -465,6 +456,56 @@ public String toString() {
...
+        }
+
+        /**
+         * Two bytes of ciphersuite id are concatenated into integer value. As an example bytes "0xC0,0xA8" will be
+         * concatenated into 0xc0a8 which in decimal notation is 49320.
+         *
+         * @return Integer number concatenated from 2 bytes.
+         */
+        public ULong getValueForSecurityObject() {
+            return ULong.valueOf(Byte.toUnsignedInt(firstByte) * 256 + Byte.toUnsignedInt(secondByte));
```
Comment (sbernard31) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:496]: Using floating point number and multiplication are maybe not best idea.

Should be better to play with binary operator like : 
- bitMask operator : `&`  
- binary notation : 0b00000
- shift operator : `<<` 

You can create unit test to check your code.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java [hunk 10]
```diff
@@ -465,6 +456,56 @@ public String toString() {
...
+        }
+
+
+        /**
+         * @return String representing hex value of concatenated two bytes.
+         */
+        @Override
+        public String toString() {
+            return Integer.toHexString(getValueForSecurityObject().intValue());
```
[the remark this answers is not in the document]
Comment (Warmek) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:505]: Here the value of ULong converted to hex would be 0xC0A8, for bytes 0xC0,0xA8, which (at least for me) seems clear enough.

[the remark this answers is not in the document]
Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapConfig.java:505]: > which (at least for me) seems clear enough.

Yep, my mistake. I read it too fast, the result is clear enough but the way it is implemented sounds a bit "winding". We have 2 bytes, we turn it in a ULong then we cast it as int then we convert it in String. 

maybe simple to just do : `String.format(".....", firstByte,secondByte)`

**Comment by sbernard31** [comment]
> We should also check that valueFromSecurityObject is a 16-bit unsigned integer.

You didn't work on it ? is it on purpose ? 

> You can create unit test to check your code.

Do you plan to create junit test about it  ? (mainly to test conversion and bad inputs)

Tell me when you reviewed your own code and consider I need to review it again. :slightly_smiling_face:

**Comment by sbernard31** [comment]
@Warmek ignore the `sortimp` [issue](https://github.com/eclipse/leshan/actions/runs/4315147451/jobs/7529109879#step:8:73), we have same problem on this PR :  https://github.com/eclipse/leshan/pull/1398#issuecomment-1451937519

I don't get it locally, I don't understand the problem for now :grimacing:

**Comment by sbernard31** [comment]
Let me know when you review your code and you satisfy about it and so it is ready for a review. 

You could do that by posting a comment or [requesting a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review).

**Review by Warmek — COMMENTED** [review]
@sbernard31 I reviewed changes and there seems to be no issues. Let me know what you think about it

**Comment by sbernard31** [comment]
@Warmek I created a PR #1408  based on yours. 

Could you review it ? :pray: If wanted you can review commit by commit this will be easier to read.

**Comment by sbernard31** [comment]
This work is included in #1408 which is now integrated in `master`.

@Warmek thx for your contribution.
